File: ios/BockMedia/Offline/OfflineDownloadIds.swift
import Foundation

extension PlayTarget {
    func downloadId() -> String {
        switch self {
        case .playlist(let id, _): return "pl-\(id)"
        case .artist(let name): return "artist-\(slug(name))"
        case .album(let name, let artist): return "album-\(slug(name))-\(slug(artist ?? ""))"
        case .song(let path, _): return "song-\(slug(path))"
        case .radio(_, let seed, let name, let path):
            switch seed {
            case .song: return "radio-song-\(slug(path ?? name))"
            case .artist: return "radio-artist-\(slug(name))"
            case .genre: return "radio-genre-\(slug(name))"
            }
        }
    }

    func downloadKindLabel() -> String {
        switch self {
        case .playlist: return "Playlist"
        case .artist: return "Artist"
        case .album: return "Album"
        case .song: return "Song"
        case .radio: return "Mix"
        }
    }
}

private func slug(_ raw: String) -> String {
    let trimmed = String(raw.trimmingCharacters(in: .whitespacesAndNewlines).prefix(120))
    return trimmed.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? trimmed
}

File: android/app/src/main/kotlin/com/bockmedia/console/local/OfflineDownloadIds.kt
package com.bockmedia.console.local

import com.bockmedia.console.domain.model.PlayTarget
import java.net.URLEncoder

fun PlayTarget.downloadId(): String = when (this) {
    is PlayTarget.Playlist -> "pl-$id"
    is PlayTarget.Artist -> "artist-${slug(name)}"
    is PlayTarget.Album -> "album-${slug(name)}-${slug(artist ?: "")}"
    is PlayTarget.Song -> "song-${slug(path)}"
    is PlayTarget.Radio -> when (seedKind) {
        PlayTarget.RadioSeedKind.Song -> "radio-song-${slug(path ?: name)}"
        PlayTarget.RadioSeedKind.Artist -> "radio-artist-${slug(name)}"
        PlayTarget.RadioSeedKind.Genre -> "radio-genre-${slug(name)}"
    }
}

fun PlayTarget.downloadKindLabel(): String = when (this) {
    is PlayTarget.Playlist -> "Playlist"
    is PlayTarget.Artist -> "Artist"
    is PlayTarget.Album -> "Album"
    is PlayTarget.Song -> "Song"
    is PlayTarget.Radio -> "Mix"
}

private fun slug(raw: String): String {
    val trimmed = raw.trim().take(120)
    return URLEncoder.encode(trimmed, "UTF-8").replace("+", "%20")
}
